### PR TITLE
Import Neighbourhood Processing and add automatic unit testing

### PR DESCRIPTION
**Pull request by benfitzpatrick**
This initial import adds the basic neighbourhood processing to IMPROVER together with inbuilt Travis continuous integration testing.

The .travis.yml file controls Travis, and is derived from the Iris one (since installing Iris and the usual associated libraries is the main problem).

The code is a straightforward shift from our internal BGD codebase, except that I have taken the liberty of reorganising it so that the `operation` method is renamed `process`, and the cube to be processed is passed in at this point rather than in the `__init__` method.

### CODE REVIEW

File: ACKNOWLEDGEMENTS.md [hunk 1]
```diff
@@ -0,0 +1,4 @@
+Credit for external pieces of work:
+
+Iris (https://github.com/scitools/iris), LGPL:
```
Comment (abooton) [inline ACKNOWLEDGEMENTS.md]: https://github.com/SciTools/iris

File: lib/improver/grids/osgb.py [hunk 2]
```diff
@@ -0,0 +1,76 @@
+# -*- coding: utf-8 -*-
```
Comment (abooton) [inline lib/improver/grids/osgb.py]: Should we have a unit test for this?

File: lib/improver/tests/test_nbhood_basic.py [hunk 3]
```diff
@@ -0,0 +1,497 @@
...
+
+
+SINGLE_POINT_RANGE_3_CENTROID = np.array([
+    [0.992, 0.968, 0.96, 0.968, 0.992],
+    [0.968, 0.944, 0.936, 0.944, 0.968],
+    [0.96, 0.936, 0.928, 0.936, 0.96],
+    [0.968, 0.944, 0.936, 0.944, 0.968],
+    [0.992, 0.968, 0.96, 0.968, 0.992]
+])
+
```
Comment (abooton) [inline lib/improver/tests/test_nbhood_basic.py]: An extra line has crept in.

File: lib/improver/tests/test_nbhood_basic.py [hunk 4]
```diff
@@ -0,0 +1,497 @@
...
+class Test_operation_radius_to_grid_cells(IrisTest):
+
+    """Test conversion of kernel radius in kilometres to grid cells."""
+
+    RADIUS_IN_KM = 6.1
+
+    def test_basic_radius_to_grid_cells(self):
+        """Test the lat-long radius-to-grid-cell conversion."""
+        cube = set_up_cube()
+        plugin = NBHood(self.RADIUS_IN_KM)
```
Comment (abooton) [inline lib/improver/tests/test_nbhood_basic.py]: Just checking, are we still continuing with the "plugin" terminology?
We are no longer using it in the main routines, but are using it in the unit-tests.

File: lib/improver/nbhood.py [hunk 5]
```diff
@@ -0,0 +1,161 @@
+# -*- coding: utf-8 -*-
```
Comment (abooton) [inline lib/improver/nbhood.py]: Longer term sort of question.. Do we have an easy way of checking this is in all new files and is correct?

File: lib/improver/nbhood.py [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+        if (grid_cells_x > self.MAX_KERNEL_CELL_RADIUS or
+                grid_cells_y > self.MAX_KERNEL_CELL_RADIUS):
+            raise ValueError(
+                ("Neighbourhood processing radius of " +
+                 "{0} km ".format(self.radius_in_km) +
+                 "exceeds maximum grid cell extent")
+            )
+        return grid_cells_x, grid_cells_y
+
+    def process(self, cube):
```
Comment (abooton) [inline lib/improver/nbhood.py]: Where are regarding the 'process' accepting a cube vs cubelist

File: .travis.yml [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+  - PREFIX=$HOME/miniconda/envs/$ENV_NAME
+
+  # Output debug info
+  - conda list
+  - conda info -a
+
+  # Setup
+  - python setup.py --with-unpack build_ext --include-dirs=${PREFIX}/include --library-dirs=${PREFIX}/lib
+  - IRIS=$(ls -d1 build/lib*/iris)
+  - mkdir $IRIS/etc
```
Comment (benfitzpatrick) [inline .travis.yml]: $IRIS already exists, so no parent creation required

File: .travis.yml [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+
+  # Set config paths
+  - SITE_CFG=$IRIS/etc/site.cfg
+  - echo "[System]" >> $SITE_CFG
```
Comment (benfitzpatrick) [inline .travis.yml]: Done

File: lib/improver/grids/osgb.py (hunk 2 above)
Comment (benfitzpatrick) [inline lib/improver/grids/osgb.py]: Not sure. It would have to basically repeat the grid information.

File: lib/improver/nbhood.py (hunk 5 above)
Comment (benfitzpatrick) [inline lib/improver/nbhood.py]: Really good point. We could add a test that loops over all Python files and checks the header. Will raise an issue.

File: lib/improver/nbhood.py (hunk 6 above)
Comment (benfitzpatrick) [inline lib/improver/nbhood.py]: I deliberately changed it to avoid getting confused with StaGE - that was the only reason.

Comment (benfitzpatrick) [inline lib/improver/nbhood.py]: I think we should aim to just always either accept a cube OR accept a cubelist. In the future, as we discussed, we could have magic decorators to handle it and type hints in Python 3 to reject invalid stuff.

File: lib/improver/nbhood.py (hunk 5 above)
Comment (benfitzpatrick) [inline lib/improver/nbhood.py]: https://github.com/metoppv/improver/issues/6

File: lib/improver/tests/test_nbhood_basic.py [hunk 9]
```diff
@@ -0,0 +1,497 @@
+# -*- coding: utf-8 -*-
```
Comment (benfitzpatrick) [inline lib/improver/tests/test_nbhood_basic.py]: Compromise: tests/unit_tests/test_nbhood_basicneighbourhoodprocessing ?

File: lib/improver/tests/test_nbhood_basic.py (hunk 3 above)
Comment (benfitzpatrick) [inline lib/improver/tests/test_nbhood_basic.py]: OK

File: lib/improver/tests/test_nbhood_basic.py (hunk 4 above)
Comment (benfitzpatrick) [inline lib/improver/tests/test_nbhood_basic.py]: True - I don't mind using it.

**Comment by benfitzpatrick** [comment]
All changes addressed, I think.

File: lib/improver/tests/test_nbhood_basic.py (hunk 9 above)
Comment (abooton) [inline lib/improver/tests/test_nbhood_basic.py]: OK.  They are camel case in StaGE, (which I think is clearer).

**Pull request merged**